transfer-sh.el

```diff
@@ -54,39 +54,76 @@
   :type '(string)
   :group 'transfer-sh)
 
+(defcustom transfer-sh-upload-agent-command
+  (cond
+   ((executable-find "curl")
+    "curl")
+   ((executable-find "wget")
+    "wget"))
+  "Command used to upload files to transfer.sh"
+  :type '(string)
+  :group 'transfer-sh)
+
+(defcustom transfer-sh-upload-agent-arguments
+  (cond
+   ((executable-find "curl")
+    (list "--silent" "--upload-file"))
+   ((executable-find "wget")
+    (list "--method" "PUT" "--output-document" "-"  "--no-verbose" "--quiet" "--body-file")))
+  "Suffix arguments to `transfer-sh-upload-agent-command'"
+  :group 'transfer-sh)
+
 ;;;###autoload
 (defun transfer-sh-upload-file-async (local-filename &optional remote-filename)
-  "Uploads file LOCAL-FILENAME to transfer.sh in background.
+  "Upload file LOCAL-FILENAME to transfer.sh in background.
 
-If no REMOTE-FILENAME is given, the LOCAL-FILENAME is used."
+If no REMOTE-FILENAME is provided, read from minibuffer the
+remote file name visible in the transfer.sh link and run
+`transfer-sh-run-upload-agent'."
   (interactive "ffile: ")
+  (or remote-filename
+      (setq remote-filename (read-from-minibuffer
+                             (format "Remote filename (default %s): "
+                                     (file-name-nondirectory local-filename))
+                             (file-name-nondirectory local-filename))))
   (async-start
    `(lambda ()
       ,(async-inject-variables "local-filename")
       ,(async-inject-variables "remote-filename")
-      (shell-command-to-string
-        (concat "curl --silent --upload-file "
-                (shell-quote-argument local-filename)
-                " " (shell-quote-argument (concat "https://transfer.sh/" remote-filename)))))
-   `(lambda (transfer-link)
-      (kill-new transfer-link)
-      (message transfer-link))))
+      ,(transfer-sh-run-upload-agent local-filename remote-filename))))
 
 ;;;###autoload
 (defun transfer-sh-upload-file (local-filename &optional remote-filename)
   "Uploads file LOCAL-FILENAME to transfer.sh.
 
-If no REMOTE-FILENAME is given, the LOCAL-FILENAME is used."
+If no REMOTE-FILENAME is provided, read from minibuffer the
+remote file name visible in the transfer.sh link and run
+`transfer-sh-run-upload-agent'."
```

Comment:
The phrasing is a bit confusing.  What about "REMOTE-FILENAME is the name used in the transfer.sh link.  If not provided, query the user".

That said, I don't think it's ideal.  In general, optional arguments should have a default value when called non-interactively and only query the user if called interactively.  You can check how a function was called with `called-interactively-p`.